<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>genomation by BIMSBbioinfo</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">genomation</h1>
        <p class="header">a toolkit to summarize, annotate and visualize genomic intervals</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/BIMSBbioinfo/genomation/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/BIMSBbioinfo/genomation/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/BIMSBbioinfo/genomation">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/BIMSBbioinfo">BIMSBbioinfo</a></p>


      </header>
      <section>
        <h1>
<a id="welcome-to-genomation" class="anchor" href="#welcome-to-genomation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Welcome to genomation</h1>

<p>This is an <a href="http://cran.r-project.org/index.html">R</a> package that contains a collection of tools for visualizing and analyzing genome-wide data sets. The package works with a variety of genomic interval file types and enables easy summarization and annotation of high throughput data sets with given genomic annotations.</p>

<h2>
<a id="features" class="anchor" href="#features" aria-hidden="true"><span class="octicon octicon-link"></span></a>Features</h2>

<ul>
<li>
<strong>Ability to work with multiple flat file formats</strong> such as <strong>BED</strong>, <strong>GFF</strong>, <strong>BAM</strong>, <strong>bigWig</strong> and <strong>generic tabular text files</strong> </li>
<li>
<strong>Annotation of genomic intervals</strong>. For example, you can see what percentage of your genomic intervals overlaps with exon/intron/promoter annotation. </li>
<li>
<strong>Summary of genomic scores or read coverages over pre-defined regions</strong>. The pre-defined regions can be equi-width such as region around TSS or could be a set of regions with varying widths such as CpG islands. This operation will result in a matrix or set of matrices containing scores for each pre-defined region </li>
<li>
<strong>Visualisation of summary matrices</strong> such as <strong>meta-region</strong>(meta-gene, meta-promoter, etc.) plots or <strong>heatmaps</strong>. 
These functions can employ K-means clustering of rows of summary matrices as well.</li>
</ul>

<p><img src="http://i.minus.com/ibqHSXWJaBcGH6.png" alt="genomation"></p>

<h2>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h2>

<p>See the package vignette <a href="http://bioconductor.org/packages/3.1/bioc/vignettes/genomation/inst/doc/GenomationManual-knitr.html">here</a>. Every function has proper documentation and examples reachable via <strong>help()</strong> function in R.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>You can install genomation via <strong>install_github()</strong> function from <a href="http://cran.r-project.org/web/packages/devtools/index.html">devtools package</a>.</p>

<div class="highlight highlight-ruby"><pre><span class="pl-c"># Install dependencies</span>
install.packages( c(<span class="pl-s1"><span class="pl-pds">"</span>data.table<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>plyr<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>reshape2<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>ggplot2<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>gridBase<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>devtools<span class="pl-pds">"</span></span>))
source(<span class="pl-s1"><span class="pl-pds">"</span>http://bioconductor.org/biocLite.R<span class="pl-pds">"</span></span>)
biocLite(c(<span class="pl-s1"><span class="pl-pds">"</span>GenomicRanges<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>rtracklayer<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>impute<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>Rsamtools<span class="pl-pds">"</span></span>))

<span class="pl-c"># install the packages</span>
library(devtools)
install_github(<span class="pl-s1"><span class="pl-pds">"</span>BIMSBbioinfo/genomation<span class="pl-pds">"</span></span>,build_vignettes<span class="pl-k">=</span><span class="pl-vo">FALSE</span>)

<span class="pl-c"># install the data package to be able to run examples in the vignette</span>
install_github(<span class="pl-s1"><span class="pl-pds">"</span>frenkiboy/genomationData<span class="pl-pds">"</span></span>,build_vignettes<span class="pl-k">=</span><span class="pl-vo">FALSE</span>)

</pre></div>

<h2>
<a id="data-import" class="anchor" href="#data-import" aria-hidden="true"><span class="octicon octicon-link"></span></a>Data import</h2>

<p>Functions such as <strong>readBed</strong>, <strong>gff2GRanges</strong>, <strong>readTranscriptFeatures</strong> and <strong>readGeneric</strong> can read multiple flat file formats as GRanges objects into R.</p>

<div class="highlight highlight-ruby"><pre><span class="pl-c"># Read a BED12 file and return a GRangesList object with promoters, exons and introns</span>
my.bed12.file <span class="pl-k">=</span> system.file(<span class="pl-s1"><span class="pl-pds">"</span>extdata/chr21.refseq.hg19.bed<span class="pl-pds">"</span></span>, package <span class="pl-k">=</span> <span class="pl-s1"><span class="pl-pds">"</span>genomation<span class="pl-pds">"</span></span>)
my.bed12.file
feats <span class="pl-k">=</span> readTranscriptFeatures(my.bed12.file)

<span class="pl-c"># Read a generic tabular text file containing genomic locations</span>
my.file<span class="pl-k">=</span>system.file(<span class="pl-s1"><span class="pl-pds">"</span>extdata<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>chr21.refseq.hg19.bed<span class="pl-pds">"</span></span>,package<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">"</span>genomation<span class="pl-pds">"</span></span>)
refseq <span class="pl-k">=</span> readGeneric(my.file,chr<span class="pl-k">=</span><span class="pl-c1">1</span>,start<span class="pl-k">=</span><span class="pl-c1">2</span>,<span class="pl-k">end</span><span class="pl-k">=</span><span class="pl-c1">3</span>,strand<span class="pl-k">=</span><span class="pl-vo">NULL</span>,
                      meta.cols<span class="pl-k">=</span>list(score<span class="pl-k">=</span><span class="pl-c1">5</span>,name<span class="pl-k">=</span><span class="pl-c1">4</span>),
                     keep.all.metadata<span class="pl-k">=</span><span class="pl-vo">FALSE</span>, zero.based<span class="pl-k">=</span><span class="pl-vo">TRUE</span>)</pre></div>

<h2>
<a id="summarize-granges-object-on-defined-regions-such-as-promoters" class="anchor" href="#summarize-granges-object-on-defined-regions-such-as-promoters" aria-hidden="true"><span class="octicon octicon-link"></span></a>Summarize GRanges object on defined regions such as promoters</h2>

<p>You can summarize GRanges objects that overlap with a set of promoters and return a ScoreMatrix object. The object will contain scores for each base in each promoter, columns correspond to bases and rows correspond to promoters.</p>

<div class="highlight highlight-ruby"><pre> data(cage)
 data(promoters)
 scores1<span class="pl-k">=</span><span class="pl-vo">ScoreMatrix</span>(target<span class="pl-k">=</span>cage,windows<span class="pl-k">=</span>promoters,strand.aware<span class="pl-k">=</span><span class="pl-vo">TRUE</span>,
                                 weight.col<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">"</span>tpm<span class="pl-pds">"</span></span>)</pre></div>

<h2>
<a id="summarize-bam-files-on-pre-defined-regions" class="anchor" href="#summarize-bam-files-on-pre-defined-regions" aria-hidden="true"><span class="octicon octicon-link"></span></a>Summarize BAM files on pre-defined regions</h2>

<p>BAM files can also be used in <strong>ScoreMatrix()</strong> function as well.</p>

<div class="highlight highlight-ruby"><pre> bam.file <span class="pl-k">=</span> system.file(<span class="pl-s1"><span class="pl-pds">'</span>tests/test.bam<span class="pl-pds">'</span></span>, package<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">'</span>genomation<span class="pl-pds">'</span></span>)
 windows <span class="pl-k">=</span> <span class="pl-vo">GRanges</span>(rep(c(<span class="pl-c1">1</span>,<span class="pl-c1">2</span>),each<span class="pl-k">=</span><span class="pl-c1">2</span>), <span class="pl-vo">IRanges</span>(rep(c(<span class="pl-c1">1</span>,<span class="pl-c1">2</span>), times<span class="pl-k">=</span><span class="pl-c1">2</span>), width<span class="pl-k">=</span><span class="pl-c1">5</span>))
 scores3 <span class="pl-k">=</span> <span class="pl-vo">ScoreMatrix</span>(target<span class="pl-k">=</span>bam.file,windows<span class="pl-k">=</span>windows, type<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">'</span>bam<span class="pl-pds">'</span></span>)</pre></div>

<h2>
<a id="summarize-bigwig-files-on-pre-defined-regions" class="anchor" href="#summarize-bigwig-files-on-pre-defined-regions" aria-hidden="true"><span class="octicon octicon-link"></span></a>Summarize BigWig files on pre-defined regions</h2>

<p>You can also use bigWig files in <strong>ScoreMatrix()</strong> function.</p>

<div class="highlight highlight-ruby"><pre>bw.file <span class="pl-k">=</span> system.file(<span class="pl-s1"><span class="pl-pds">'</span>tests/test.bw<span class="pl-pds">'</span></span>, package<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">'</span>rtracklayer<span class="pl-pds">'</span></span>)
windows <span class="pl-k">=</span> <span class="pl-vo">GRanges</span>(rep(<span class="pl-s1"><span class="pl-pds">'</span>chr2<span class="pl-pds">'</span></span>,each<span class="pl-k">=</span><span class="pl-c1">4</span>), <span class="pl-vo">IRanges</span>(start<span class="pl-k">=</span>c(<span class="pl-c1">250</span>,<span class="pl-c1">350</span>,<span class="pl-c1">450</span>,<span class="pl-c1">550</span>), width<span class="pl-k">=</span><span class="pl-c1">50</span>))
scores3 <span class="pl-k">=</span> <span class="pl-vo">ScoreMatrix</span>(target<span class="pl-k">=</span>bw.file ,windows<span class="pl-k">=</span>windows, type<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">'</span>bigWig<span class="pl-pds">'</span></span>)</pre></div>

<h2>
<a id="visualize-summary-matrices-as-heatmap" class="anchor" href="#visualize-summary-matrices-as-heatmap" aria-hidden="true"><span class="octicon octicon-link"></span></a>Visualize summary matrices as heatmap</h2>

<p>ScoreMatrix or ScoreMatrixList objects can be visualized with <strong>heatMatrix, multiHeatMatrix, plotMeta and heatMeta</strong> functions.</p>

<div class="highlight highlight-ruby"><pre>data(cage)
data(promoters)
scores1<span class="pl-k">=</span><span class="pl-vo">ScoreMatrix</span>(target<span class="pl-k">=</span>cage,windows<span class="pl-k">=</span>promoters,strand.aware<span class="pl-k">=</span><span class="pl-vo">TRUE</span>)

data(cpgi)
scores2<span class="pl-k">=</span><span class="pl-vo">ScoreMatrix</span>(target<span class="pl-k">=</span>cpgi,windows<span class="pl-k">=</span>promoters,strand.aware<span class="pl-k">=</span><span class="pl-vo">TRUE</span>)

sml<span class="pl-k">=</span><span class="pl-k">new</span>(<span class="pl-s1"><span class="pl-pds">"</span>ScoreMatrixList<span class="pl-pds">"</span></span>,list(a<span class="pl-k">=</span>scores1,b<span class="pl-k">=</span>scores2))
multiHeatMatrix(sml,kmeans<span class="pl-k">=</span><span class="pl-vo">TRUE</span>,k<span class="pl-k">=</span><span class="pl-c1">2</span>,matrix.main<span class="pl-k">=</span>c(<span class="pl-s1"><span class="pl-pds">"</span>cage<span class="pl-pds">"</span></span>,<span class="pl-s1"><span class="pl-pds">"</span>CpGi<span class="pl-pds">"</span></span>),cex.axis<span class="pl-k">=</span><span class="pl-c1">0.8</span>)
</pre></div>

<h2>
<a id="visualize-summary-matrices-as-meta-region-plots" class="anchor" href="#visualize-summary-matrices-as-meta-region-plots" aria-hidden="true"><span class="octicon octicon-link"></span></a>Visualize summary matrices as meta-region plots</h2>

<div class="highlight highlight-ruby"><pre>plotMeta(mat<span class="pl-k">=</span>sml,overlay<span class="pl-k">=</span><span class="pl-vo">TRUE</span>,main<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">"</span>my plotowski<span class="pl-pds">"</span></span>)
heatMeta(mat<span class="pl-k">=</span>sml,main<span class="pl-k">=</span><span class="pl-s1"><span class="pl-pds">"</span>my plotowski<span class="pl-pds">"</span></span>)</pre></div>

<h1>
<a id="authors-and-contributors" class="anchor" href="#authors-and-contributors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors and Contributors</h1>

<p>Vedran Franke (<a href="https://github.com/frenkiboy" class="user-mention">@frenkiboy</a>) and Altuna Akalin (<a href="https://github.com/al2na" class="user-mention">@al2na</a>) authored this package. You can contribute by checking out the "development" branch, making changes and submitting a pull request.</p>

<h1>
<a id="support-or-contact" class="anchor" href="#support-or-contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Support or Contact</h1>

<p>send an e-mail to <a href="mailto:genomation@googlegroups.com">genomation@googlegroups.com</a> or use the web interface
to post a question
<a href="https://groups.google.com/forum/#!forum/genomation">https://groups.google.com/forum/#!forum/genomation</a></p>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-47762010-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
