<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>genomation by BIMSBbioinfo</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">genomation</h1>
      <h2 class="project-tagline">a toolkit to summarize, annotate and visualize genomic intervals</h2>
      <a href="https://github.com/BIMSBbioinfo/genomation" class="btn">View on GitHub</a>
      <a href="https://github.com/BIMSBbioinfo/genomation/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/BIMSBbioinfo/genomation/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="welcome-to-genomation" class="anchor" href="#welcome-to-genomation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Welcome to genomation</h1>

<p>This is an <a href="http://cran.r-project.org/index.html">R</a> package that contains a collection of tools for visualizing and analyzing genome-wide data sets. The package works with a variety of genomic interval file types and enables easy summarization and annotation of high throughput data sets with given genomic annotations.</p>

<h2>
<a id="features" class="anchor" href="#features" aria-hidden="true"><span class="octicon octicon-link"></span></a>Features</h2>

<ul>
<li>
<strong>Ability to work with multiple flat file formats</strong> such as <strong>BED</strong>, <strong>GFF</strong>, <strong>BAM</strong>, <strong>bigWig</strong> and <strong>generic tabular text files</strong> </li>
<li>
<strong>Annotation of genomic intervals</strong>. For example, you can see what percentage of your genomic intervals overlaps with exon/intron/promoter annotation. </li>
<li>
<strong>Summary of genomic scores or read coverages over pre-defined regions</strong>. The pre-defined regions can be equi-width such as region around TSS or could be a set of regions with varying widths such as CpG islands. This operation will result in a matrix or set of matrices containing scores for each pre-defined region </li>
<li>
<strong>Visualisation of summary matrices</strong> such as <strong>meta-region</strong>(meta-gene, meta-promoter, etc.) plots or <strong>heatmaps</strong>. 
These functions can employ K-means clustering of rows of summary matrices as well.</li>
</ul>

<p><img src="http://i.minus.com/ibqHSXWJaBcGH6.png" alt="genomation"></p>

<h2>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h2>

<ul>
<li><p>See the package vignette <a href="http://bioconductor.org/packages/3.1/bioc/vignettes/genomation/inst/doc/GenomationManual-knitr.html">here</a>. Every function has proper documentation and examples reachable via <strong>help()</strong> function in R.</p></li>
<li><p>See blogposts on <strong>genomation</strong> <a href="http://zvfak.blogspot.com/search/label/genomation">http://zvfak.blogspot.com/search/label/genomation</a></p></li>
</ul>

<h2>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Citation</h2>

<p>Akalin A, Franke V, Vlahoviček K, Mason CE, Schübeler D. <a href="http://www.ncbi.nlm.nih.gov/pubmed/25417204">genomation: a toolkit to summarize, annotate and visualize genomic intervals</a>. Bioinformatics. 2014 Nov 21. pii: btu775</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<h4>
<a id="install-via-bioconductor" class="anchor" href="#install-via-bioconductor" aria-hidden="true"><span class="octicon octicon-link"></span></a>Install via Bioconductor</h4>

<div class="highlight highlight-r"><pre>source(<span class="pl-s"><span class="pl-pds">"</span>http://bioconductor.org/biocLite.R<span class="pl-pds">"</span></span>)
biocLite(<span class="pl-s"><span class="pl-pds">"</span>genomation<span class="pl-pds">"</span></span>)</pre></div>

<h4>
<a id="install-the-latest-version-via-devtoolsinstall_github" class="anchor" href="#install-the-latest-version-via-devtoolsinstall_github" aria-hidden="true"><span class="octicon octicon-link"></span></a>Install the latest version via devtools::install_github</h4>

<p>You can install genomation via <strong>install_github()</strong> function from <a href="http://cran.r-project.org/web/packages/devtools/index.html">devtools package</a>.</p>

<div class="highlight highlight-r"><pre><span class="pl-c"># Install dependencies</span>
install.packages( c(<span class="pl-s"><span class="pl-pds">"</span>data.table<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>plyr<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>reshape2<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>ggplot2<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>gridBase<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>devtools<span class="pl-pds">"</span></span>))
source(<span class="pl-s"><span class="pl-pds">"</span>http://bioconductor.org/biocLite.R<span class="pl-pds">"</span></span>)
biocLite(c(<span class="pl-s"><span class="pl-pds">"</span>GenomicRanges<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>rtracklayer<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>impute<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>Rsamtools<span class="pl-pds">"</span></span>))

<span class="pl-c"># install the packages</span>
library(<span class="pl-smi">devtools</span>)
install_github(<span class="pl-s"><span class="pl-pds">"</span>BIMSBbioinfo/genomation<span class="pl-pds">"</span></span>,<span class="pl-v">build_vignettes</span><span class="pl-k">=</span><span class="pl-c1">FALSE</span>)

<span class="pl-c"># install the data package to be able to run examples in the vignette</span>
install_github(<span class="pl-s"><span class="pl-pds">"</span>frenkiboy/genomationData<span class="pl-pds">"</span></span>,<span class="pl-v">build_vignettes</span><span class="pl-k">=</span><span class="pl-c1">FALSE</span>)

</pre></div>

<h2>
<a id="data-import" class="anchor" href="#data-import" aria-hidden="true"><span class="octicon octicon-link"></span></a>Data import</h2>

<p>Functions such as <strong>readBed</strong>, <strong>gff2GRanges</strong>, <strong>readTranscriptFeatures</strong> and <strong>readGeneric</strong> can read multiple flat file formats as GRanges objects into R.</p>

<div class="highlight highlight-r"><pre><span class="pl-c"># Read a BED12 file and return a GRangesList object with promoters, exons and introns</span>
<span class="pl-v">my.bed12.file</span> <span class="pl-k">=</span> system.file(<span class="pl-s"><span class="pl-pds">"</span>extdata/chr21.refseq.hg19.bed<span class="pl-pds">"</span></span>, <span class="pl-v">package</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>genomation<span class="pl-pds">"</span></span>)
<span class="pl-smi">my.bed12.file</span>
<span class="pl-v">feats</span> <span class="pl-k">=</span> readTranscriptFeatures(<span class="pl-smi">my.bed12.file</span>)

<span class="pl-c"># Read a generic tabular text file containing genomic locations</span>
<span class="pl-v">my.file</span><span class="pl-k">=</span>system.file(<span class="pl-s"><span class="pl-pds">"</span>extdata<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>chr21.refseq.hg19.bed<span class="pl-pds">"</span></span>,<span class="pl-v">package</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>genomation<span class="pl-pds">"</span></span>)
<span class="pl-v">refseq</span> <span class="pl-k">=</span> readGeneric(<span class="pl-smi">my.file</span>,<span class="pl-v">chr</span><span class="pl-k">=</span><span class="pl-c1">1</span>,<span class="pl-v">start</span><span class="pl-k">=</span><span class="pl-c1">2</span>,<span class="pl-v">end</span><span class="pl-k">=</span><span class="pl-c1">3</span>,<span class="pl-v">strand</span><span class="pl-k">=</span><span class="pl-c1">NULL</span>,
                      <span class="pl-v">meta.cols</span><span class="pl-k">=</span><span class="pl-k">list</span>(<span class="pl-v">score</span><span class="pl-k">=</span><span class="pl-c1">5</span>,<span class="pl-v">name</span><span class="pl-k">=</span><span class="pl-c1">4</span>),
                     <span class="pl-v">keep.all.metadata</span><span class="pl-k">=</span><span class="pl-c1">FALSE</span>, <span class="pl-v">zero.based</span><span class="pl-k">=</span><span class="pl-c1">TRUE</span>)</pre></div>

<h2>
<a id="summarize-granges-object-on-defined-regions-such-as-promoters" class="anchor" href="#summarize-granges-object-on-defined-regions-such-as-promoters" aria-hidden="true"><span class="octicon octicon-link"></span></a>Summarize GRanges object on defined regions such as promoters</h2>

<p>You can summarize GRanges objects that overlap with a set of promoters and return a ScoreMatrix object. The object will contain scores for each base in each promoter, columns correspond to bases and rows correspond to promoters.</p>

<div class="highlight highlight-r"><pre> data(<span class="pl-smi">cage</span>)
 data(<span class="pl-smi">promoters</span>)
 <span class="pl-v">scores1</span><span class="pl-k">=</span>ScoreMatrix(<span class="pl-v">target</span><span class="pl-k">=</span><span class="pl-smi">cage</span>,<span class="pl-v">windows</span><span class="pl-k">=</span><span class="pl-smi">promoters</span>,<span class="pl-v">strand.aware</span><span class="pl-k">=</span><span class="pl-c1">TRUE</span>,
                                 <span class="pl-v">weight.col</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>tpm<span class="pl-pds">"</span></span>)</pre></div>

<h2>
<a id="summarize-bam-files-on-pre-defined-regions" class="anchor" href="#summarize-bam-files-on-pre-defined-regions" aria-hidden="true"><span class="octicon octicon-link"></span></a>Summarize BAM files on pre-defined regions</h2>

<p>BAM files can also be used in <strong>ScoreMatrix()</strong> function as well.</p>

<div class="highlight highlight-r"><pre> <span class="pl-v">bam.file</span> <span class="pl-k">=</span> system.file(<span class="pl-s"><span class="pl-pds">'</span>tests/test.bam<span class="pl-pds">'</span></span>, <span class="pl-v">package</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>genomation<span class="pl-pds">'</span></span>)
 <span class="pl-v">windows</span> <span class="pl-k">=</span> GRanges(rep(c(<span class="pl-c1">1</span>,<span class="pl-c1">2</span>),<span class="pl-v">each</span><span class="pl-k">=</span><span class="pl-c1">2</span>), IRanges(rep(c(<span class="pl-c1">1</span>,<span class="pl-c1">2</span>), <span class="pl-v">times</span><span class="pl-k">=</span><span class="pl-c1">2</span>), <span class="pl-v">width</span><span class="pl-k">=</span><span class="pl-c1">5</span>))
 <span class="pl-v">scores3</span> <span class="pl-k">=</span> ScoreMatrix(<span class="pl-v">target</span><span class="pl-k">=</span><span class="pl-smi">bam.file</span>,<span class="pl-v">windows</span><span class="pl-k">=</span><span class="pl-smi">windows</span>, <span class="pl-v">type</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>bam<span class="pl-pds">'</span></span>)</pre></div>

<h2>
<a id="summarize-bigwig-files-on-pre-defined-regions" class="anchor" href="#summarize-bigwig-files-on-pre-defined-regions" aria-hidden="true"><span class="octicon octicon-link"></span></a>Summarize BigWig files on pre-defined regions</h2>

<p>You can also use bigWig files in <strong>ScoreMatrix()</strong> function.</p>

<div class="highlight highlight-r"><pre><span class="pl-v">bw.file</span> <span class="pl-k">=</span> system.file(<span class="pl-s"><span class="pl-pds">'</span>tests/test.bw<span class="pl-pds">'</span></span>, <span class="pl-v">package</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>rtracklayer<span class="pl-pds">'</span></span>)
<span class="pl-v">windows</span> <span class="pl-k">=</span> GRanges(rep(<span class="pl-s"><span class="pl-pds">'</span>chr2<span class="pl-pds">'</span></span>,<span class="pl-v">each</span><span class="pl-k">=</span><span class="pl-c1">4</span>), IRanges(<span class="pl-v">start</span><span class="pl-k">=</span>c(<span class="pl-c1">250</span>,<span class="pl-c1">350</span>,<span class="pl-c1">450</span>,<span class="pl-c1">550</span>), <span class="pl-v">width</span><span class="pl-k">=</span><span class="pl-c1">50</span>))
<span class="pl-v">scores3</span> <span class="pl-k">=</span> ScoreMatrix(<span class="pl-v">target</span><span class="pl-k">=</span><span class="pl-smi">bw.file</span> ,<span class="pl-v">windows</span><span class="pl-k">=</span><span class="pl-smi">windows</span>, <span class="pl-v">type</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>bigWig<span class="pl-pds">'</span></span>)</pre></div>

<h2>
<a id="visualize-summary-matrices-as-heatmap" class="anchor" href="#visualize-summary-matrices-as-heatmap" aria-hidden="true"><span class="octicon octicon-link"></span></a>Visualize summary matrices as heatmap</h2>

<p>ScoreMatrix or ScoreMatrixList objects can be visualized with <strong>heatMatrix, multiHeatMatrix, plotMeta and heatMeta</strong> functions.</p>

<div class="highlight highlight-r"><pre>data(<span class="pl-smi">cage</span>)
data(<span class="pl-smi">promoters</span>)
<span class="pl-v">scores1</span><span class="pl-k">=</span>ScoreMatrix(<span class="pl-v">target</span><span class="pl-k">=</span><span class="pl-smi">cage</span>,<span class="pl-v">windows</span><span class="pl-k">=</span><span class="pl-smi">promoters</span>,<span class="pl-v">strand.aware</span><span class="pl-k">=</span><span class="pl-c1">TRUE</span>)

data(<span class="pl-smi">cpgi</span>)
<span class="pl-v">scores2</span><span class="pl-k">=</span>ScoreMatrix(<span class="pl-v">target</span><span class="pl-k">=</span><span class="pl-smi">cpgi</span>,<span class="pl-v">windows</span><span class="pl-k">=</span><span class="pl-smi">promoters</span>,<span class="pl-v">strand.aware</span><span class="pl-k">=</span><span class="pl-c1">TRUE</span>)

<span class="pl-v">sml</span><span class="pl-k">=</span>new(<span class="pl-s"><span class="pl-pds">"</span>ScoreMatrixList<span class="pl-pds">"</span></span>,<span class="pl-k">list</span>(<span class="pl-v">a</span><span class="pl-k">=</span><span class="pl-smi">scores1</span>,<span class="pl-v">b</span><span class="pl-k">=</span><span class="pl-smi">scores2</span>))
multiHeatMatrix(<span class="pl-smi">sml</span>,<span class="pl-v">kmeans</span><span class="pl-k">=</span><span class="pl-c1">TRUE</span>,<span class="pl-v">k</span><span class="pl-k">=</span><span class="pl-c1">2</span>,<span class="pl-v">matrix.main</span><span class="pl-k">=</span>c(<span class="pl-s"><span class="pl-pds">"</span>cage<span class="pl-pds">"</span></span>,<span class="pl-s"><span class="pl-pds">"</span>CpGi<span class="pl-pds">"</span></span>),<span class="pl-v">cex.axis</span><span class="pl-k">=</span><span class="pl-c1">0.8</span>)
</pre></div>

<h2>
<a id="visualize-summary-matrices-as-meta-region-plots" class="anchor" href="#visualize-summary-matrices-as-meta-region-plots" aria-hidden="true"><span class="octicon octicon-link"></span></a>Visualize summary matrices as meta-region plots</h2>

<div class="highlight highlight-r"><pre>plotMeta(<span class="pl-v">mat</span><span class="pl-k">=</span><span class="pl-smi">sml</span>,<span class="pl-v">overlay</span><span class="pl-k">=</span><span class="pl-c1">TRUE</span>,<span class="pl-v">main</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>my plotowski<span class="pl-pds">"</span></span>)
heatMeta(<span class="pl-v">mat</span><span class="pl-k">=</span><span class="pl-smi">sml</span>,<span class="pl-v">main</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>my plotowski<span class="pl-pds">"</span></span>)</pre></div>

<h1>
<a id="authors-and-contributors" class="anchor" href="#authors-and-contributors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors and Contributors</h1>

<p>Vedran Franke (<a href="https://github.com/frenkiboy" class="user-mention">@frenkiboy</a>) and Altuna Akalin (<a href="https://github.com/al2na" class="user-mention">@al2na</a>) initially authored this package. Check <a href="https://github.com/BIMSBbioinfo/genomation/graphs/contributors">here</a> to see other contributors . You can contribute by checking out the "development" branch, making changes and submitting a pull request.</p>

<h1>
<a id="support-or-contact" class="anchor" href="#support-or-contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Support or Contact</h1>

<p>send an e-mail to <a href="mailto:genomation@googlegroups.com">genomation@googlegroups.com</a> or use the web interface
to post a question
<a href="https://groups.google.com/forum/#!forum/genomation">https://groups.google.com/forum/#!forum/genomation</a></p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/BIMSBbioinfo/genomation">genomation</a> is maintained by <a href="https://github.com/BIMSBbioinfo">BIMSBbioinfo</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-47762010-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>

