<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>genomation by al2na</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">genomation</h1>
        <p class="header">a toolkit to summarize, annotate and visualize genomic intervals</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/al2na/genomation/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/al2na/genomation/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/al2na/genomation">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/al2na">al2na</a></p>


      </header>
      <section>
        <h1>
<a name="welcome-to-genomation" class="anchor" href="#welcome-to-genomation"><span class="octicon octicon-link"></span></a>Welcome to genomation</h1>

<p>This is an <a href="http://cran.r-project.org/index.html">R</a> package that contains a collection of tools for visualizing and analyzing genome-wide data sets. The package works with a variety of genomic interval file types and enables easy summarization and annotation of high throughput data sets with given genomic annotations.</p>

<h2>
<a name="features" class="anchor" href="#features"><span class="octicon octicon-link"></span></a>Features</h2>

<ul>
<li>
<strong>Ability to work with multiple flat file formats</strong> such as <strong>BED</strong>, <strong>GFF</strong>, <strong>BAM</strong>, <strong>bigWig</strong> and <strong>generic tabular text files</strong> </li>
<li>
<strong>Annotation of genomic intervals</strong>. For example, you can see what percentage of your genomic intervals overlaps with exon/intron/promoter annotation. </li>
<li>
<strong>Summary of genomic scores or read coverages over pre-defined regions</strong>. The pre-defined regions can be equi-width such as region around TSS or could be a set of regions with varying widths such as CpG islands. This operation will result in a matrix or set of matrices containing scores for each pre-defined region </li>
<li>
<strong>Visualisation of summary matrices</strong> such as <strong>meta-region</strong>(meta-gene, meta-promoter, etc.) plots or <strong>heatmaps</strong>. 
These functions can employ K-means clustering of rows of summary matrices as well.</li>
</ul><p><img src="http://i.minus.com/ibqHSXWJaBcGH6.png" alt="genomation"></p>

<h2>
<a name="documentation" class="anchor" href="#documentation"><span class="octicon octicon-link"></span></a>Documentation</h2>

<p>See the package vignette <a href="https://github.com/al2na/genomation/raw/master/inst/doc/GenomationManual-knitr.pdf">here</a>. Every function has proper documentation and examples reachable via <strong>help()</strong> function in R.</p>

<h2>
<a name="installation" class="anchor" href="#installation"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>You can install genomation via <strong>install_github()</strong> function from <a href="http://cran.r-project.org/web/packages/devtools/index.html">devtools package</a>.</p>

<div class="highlight highlight-ruby"><pre><span class="c1"># Install dependencies</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span> <span class="n">c</span><span class="p">(</span><span class="s2">"data.table"</span><span class="p">,</span><span class="s2">"plyr"</span><span class="p">,</span><span class="s2">"reshape2"</span><span class="p">,</span><span class="s2">"ggplot2"</span><span class="p">,</span><span class="s2">"gridBase"</span><span class="p">,</span><span class="s2">"devtools"</span><span class="p">))</span>
<span class="n">source</span><span class="p">(</span><span class="s2">"http://bioconductor.org/biocLite.R"</span><span class="p">)</span>
<span class="n">biocLite</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"GenomicRanges"</span><span class="p">,</span><span class="s2">"rtracklayer"</span><span class="p">,</span><span class="s2">"impute"</span><span class="p">,</span><span class="s2">"Rsamtools"</span><span class="p">))</span>

<span class="c1"># install the packages</span>
<span class="n">library</span><span class="p">(</span><span class="n">devtools</span><span class="p">)</span>
<span class="n">install_github</span><span class="p">(</span><span class="s2">"genomation"</span><span class="p">,</span> <span class="n">username</span> <span class="o">=</span> <span class="s2">"al2na"</span><span class="p">,</span><span class="n">build_vignettes</span><span class="o">=</span><span class="no">FALSE</span><span class="p">)</span>
</pre></div>

<h2>
<a name="data-import" class="anchor" href="#data-import"><span class="octicon octicon-link"></span></a>Data import</h2>

<p>Functions such as <strong>readBed</strong>, <strong>gff2GRanges</strong>, <strong>readTranscriptFeatures</strong> and <strong>readGeneric</strong> can read multiple flat file formats as GRanges objects into R.</p>

<div class="highlight highlight-ruby"><pre><span class="c1"># Read a BED12 file and return a GRangesList object with promoters, exons and introns</span>
<span class="n">my</span><span class="o">.</span><span class="n">bed12</span><span class="o">.</span><span class="n">file</span> <span class="o">=</span> <span class="nb">system</span><span class="o">.</span><span class="n">file</span><span class="p">(</span><span class="s2">"extdata/chr21.refseq.hg19.bed"</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s2">"genomation"</span><span class="p">)</span>
<span class="n">my</span><span class="o">.</span><span class="n">bed12</span><span class="o">.</span><span class="n">file</span>
<span class="n">feats</span> <span class="o">=</span> <span class="n">readTranscriptFeatures</span><span class="p">(</span><span class="n">my</span><span class="o">.</span><span class="n">bed12</span><span class="o">.</span><span class="n">file</span><span class="p">)</span>

<span class="c1"># Read a generic tabular text file containing genomic locations</span>
<span class="n">my</span><span class="o">.</span><span class="n">file</span><span class="o">=</span><span class="nb">system</span><span class="o">.</span><span class="n">file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="s2">"chr21.refseq.hg19.bed"</span><span class="p">,</span><span class="n">package</span><span class="o">=</span><span class="s2">"genomation"</span><span class="p">)</span>
<span class="n">refseq</span> <span class="o">=</span> <span class="n">readGeneric</span><span class="p">(</span><span class="n">my</span><span class="o">.</span><span class="n">file</span><span class="p">,</span><span class="n">chr</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">start</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="k">end</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">strand</span><span class="o">=</span><span class="no">NULL</span><span class="p">,</span>
                      <span class="n">meta</span><span class="o">.</span><span class="n">cols</span><span class="o">=</span><span class="n">list</span><span class="p">(</span><span class="n">score</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="nb">name</span><span class="o">=</span><span class="mi">4</span><span class="p">),</span>
                     <span class="n">keep</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">metadata</span><span class="o">=</span><span class="no">FALSE</span><span class="p">,</span> <span class="n">zero</span><span class="o">.</span><span class="n">based</span><span class="o">=</span><span class="no">TRUE</span><span class="p">)</span>
</pre></div>

<h2>
<a name="summarize-granges-object-on-defined-regions-such-as-promoters" class="anchor" href="#summarize-granges-object-on-defined-regions-such-as-promoters"><span class="octicon octicon-link"></span></a>Summarize GRanges object on defined regions such as promoters</h2>

<p>You can summarize GRanges objects that overlap with a set of promoters and return a ScoreMatrix object. The object will contain scores for each base in each promoter, columns correspond to bases and rows correspond to promoters.</p>

<div class="highlight highlight-ruby"><pre> <span class="n">data</span><span class="p">(</span><span class="n">cage</span><span class="p">)</span>
 <span class="n">data</span><span class="p">(</span><span class="n">promoters</span><span class="p">)</span>
 <span class="n">scores1</span><span class="o">=</span><span class="no">ScoreMatrix</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">cage</span><span class="p">,</span><span class="n">windows</span><span class="o">=</span><span class="n">promoters</span><span class="p">,</span><span class="n">strand</span><span class="o">.</span><span class="n">aware</span><span class="o">=</span><span class="no">TRUE</span><span class="p">,</span>
                                 <span class="n">weight</span><span class="o">.</span><span class="n">col</span><span class="o">=</span><span class="s2">"tpm"</span><span class="p">)</span>
</pre></div>

<h2>
<a name="summarize-bam-files-on-pre-defined-regions" class="anchor" href="#summarize-bam-files-on-pre-defined-regions"><span class="octicon octicon-link"></span></a>Summarize BAM files on pre-defined regions</h2>

<p>BAM files can also be used in <strong>ScoreMatrix()</strong> function as well.</p>

<div class="highlight highlight-ruby"><pre> <span class="n">bam</span><span class="o">.</span><span class="n">file</span> <span class="o">=</span> <span class="nb">system</span><span class="o">.</span><span class="n">file</span><span class="p">(</span><span class="s1">'tests/test.bam'</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s1">'genomation'</span><span class="p">)</span>
 <span class="n">windows</span> <span class="o">=</span> <span class="no">GRanges</span><span class="p">(</span><span class="n">rep</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span><span class="n">each</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span> <span class="no">IRanges</span><span class="p">(</span><span class="n">rep</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">times</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span> <span class="n">width</span><span class="o">=</span><span class="mi">5</span><span class="p">))</span>
 <span class="n">scores3</span> <span class="o">=</span> <span class="no">ScoreMatrix</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">bam</span><span class="o">.</span><span class="n">file</span><span class="p">,</span><span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s1">'bam'</span><span class="p">)</span>
</pre></div>

<h2>
<a name="summarize-bigwig-files-on-pre-defined-regions" class="anchor" href="#summarize-bigwig-files-on-pre-defined-regions"><span class="octicon octicon-link"></span></a>Summarize BigWig files on pre-defined regions</h2>

<p>You can also use bigWig files in <strong>ScoreMatrix()</strong> function.</p>

<div class="highlight highlight-ruby"><pre><span class="n">bw</span><span class="o">.</span><span class="n">file</span> <span class="o">=</span> <span class="nb">system</span><span class="o">.</span><span class="n">file</span><span class="p">(</span><span class="s1">'tests/test.bw'</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s1">'rtracklayer'</span><span class="p">)</span>
<span class="n">windows</span> <span class="o">=</span> <span class="no">GRanges</span><span class="p">(</span><span class="n">rep</span><span class="p">(</span><span class="s1">'chr2'</span><span class="p">,</span><span class="n">each</span><span class="o">=</span><span class="mi">4</span><span class="p">),</span> <span class="no">IRanges</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="mi">250</span><span class="p">,</span><span class="mi">350</span><span class="p">,</span><span class="mi">450</span><span class="p">,</span><span class="mi">550</span><span class="p">),</span> <span class="n">width</span><span class="o">=</span><span class="mi">50</span><span class="p">))</span>
<span class="n">scores3</span> <span class="o">=</span> <span class="no">ScoreMatrix</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">bw</span><span class="o">.</span><span class="n">file</span> <span class="p">,</span><span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s1">'bigWig'</span><span class="p">)</span>
</pre></div>

<h2>
<a name="visualize-summary-matrices-as-heatmap" class="anchor" href="#visualize-summary-matrices-as-heatmap"><span class="octicon octicon-link"></span></a>Visualize summary matrices as heatmap</h2>

<p>ScoreMatrix or ScoreMatrixList objects can be visualized with <strong>heatMatrix, multiHeatMatrix, plotMeta and heatMeta</strong> functions.</p>

<div class="highlight highlight-ruby"><pre><span class="n">data</span><span class="p">(</span><span class="n">cage</span><span class="p">)</span>
<span class="n">data</span><span class="p">(</span><span class="n">promoters</span><span class="p">)</span>
<span class="n">scores1</span><span class="o">=</span><span class="no">ScoreMatrix</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">cage</span><span class="p">,</span><span class="n">windows</span><span class="o">=</span><span class="n">promoters</span><span class="p">,</span><span class="n">strand</span><span class="o">.</span><span class="n">aware</span><span class="o">=</span><span class="no">TRUE</span><span class="p">)</span>

<span class="n">data</span><span class="p">(</span><span class="n">cpgi</span><span class="p">)</span>
<span class="n">scores2</span><span class="o">=</span><span class="no">ScoreMatrix</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">cpgi</span><span class="p">,</span><span class="n">windows</span><span class="o">=</span><span class="n">promoters</span><span class="p">,</span><span class="n">strand</span><span class="o">.</span><span class="n">aware</span><span class="o">=</span><span class="no">TRUE</span><span class="p">)</span>

<span class="n">sml</span><span class="o">=</span><span class="kp">new</span><span class="p">(</span><span class="s2">"ScoreMatrixList"</span><span class="p">,</span><span class="n">list</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">scores1</span><span class="p">,</span><span class="n">b</span><span class="o">=</span><span class="n">scores2</span><span class="p">))</span>
<span class="n">multiHeatMatrix</span><span class="p">(</span><span class="n">sml</span><span class="p">,</span><span class="n">kmeans</span><span class="o">=</span><span class="no">TRUE</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">matrix</span><span class="o">.</span><span class="n">main</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">"cage"</span><span class="p">,</span><span class="s2">"CpGi"</span><span class="p">),</span><span class="n">cex</span><span class="o">.</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="o">.</span><span class="mi">8</span><span class="p">)</span>

</pre></div>

<h2>
<a name="visualize-summary-matrices-as-meta-region-plots" class="anchor" href="#visualize-summary-matrices-as-meta-region-plots"><span class="octicon octicon-link"></span></a>Visualize summary matrices as meta-region plots</h2>

<div class="highlight highlight-ruby"><pre><span class="n">plotMeta</span><span class="p">(</span><span class="n">mat</span><span class="o">=</span><span class="n">sml</span><span class="p">,</span><span class="n">overlay</span><span class="o">=</span><span class="no">TRUE</span><span class="p">,</span><span class="n">main</span><span class="o">=</span><span class="s2">"my plotowski"</span><span class="p">)</span>
<span class="n">heatMeta</span><span class="p">(</span><span class="n">mat</span><span class="o">=</span><span class="n">sml</span><span class="p">,</span><span class="n">main</span><span class="o">=</span><span class="s2">"my plotowski"</span><span class="p">)</span>
</pre></div>

<h1>
<a name="authors-and-contributors" class="anchor" href="#authors-and-contributors"><span class="octicon octicon-link"></span></a>Authors and Contributors</h1>

<p>Vedran Franke (<a href="https://github.com/frenkiboy" class="user-mention">@frenkiboy</a>) and Altuna Akalin (<a href="https://github.com/al2na" class="user-mention">@al2na</a>) authored this package. You can contribute by checking out the "development" branch, making changes and submitting a pull request.</p>

<h1>
<a name="support-or-contact" class="anchor" href="#support-or-contact"><span class="octicon octicon-link"></span></a>Support or Contact</h1>

<p>send an e-mail to <a href="mailto:genomation@googlegroups.com">genomation@googlegroups.com</a> or use the web interface
to post a question
<a href="https://groups.google.com/forum/#!forum/genomation">https://groups.google.com/forum/#!forum/genomation</a></p>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-47762010-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
