%\VignetteKeywords{genomation}
%\VignettePackage{genomation}
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{genomation: User Guide}
% !Rnw weave = knitr
\documentclass{article}
\usepackage{geometry}
\usepackage{wrapfig}

\geometry{verbose,tmargin=2.5cm,bmargin=2.5cm,lmargin=2.5cm,rmargin=2.5cm}


\title{genomation - a toolkit for annotation and visualization of genomic data}

\author{Altuna Akalin\\ \texttt{altuna.akalin@fmi.ch}\\
\and
Vedran Franke \\ \texttt{vedran.franke@gmail.com} }




\newcommand{\Rpackage}[1]{{\textit{#1}}}



<<setup2, include=FALSE, cache=FALSE>>=
library(knitr)
options(replace.assign=TRUE,width=50)
opts_chunk$set(root.dir=getwd(),
               fig.path='inst/doc/Figures', 
               cache.path='cache/graphics-', 
               fig.align='center', 
               dev='pdf', 
               fig.width=5, 
               fig.height=5, 
               fig.show='hold', 
               cache=FALSE,
               par=TRUE)
knit_hooks$set(par=function(before, options, envir){
if (before && options$fig.show!='none') par(mar=c(4,4,.1,.1),cex.lab=.95,cex.axis=.9,mgp=c(2,.7,0),tcl=-.3)
}, crop=hook_pdfcrop)
@

\begin{document}
\maketitle
\tableofcontents

% ----------------------------------------------------------------- %
\section{Introduction}

Recent advances in sequencing technologies have enabled a downpoor of biological data. The sheer amount of data has impeeded the extraction of useful knowledge and novel hypothesis generation.
\begin{wrapfigure}{r}{0.4\textwidth}
 \vspace{-20pt}
  \begin{center}
    \includegraphics[width=0.4\textwidth]{Figures/genomationFlowChart1.pdf}
  \end{center}
  \caption{Bulk visualization for different genomic feature datasets flowchart}
   \vspace{-20pt}
\end{wrapfigure} 
\Rpackage{genomation} is a toolkit for annotation and in bulk visualization of 
genomic features (scored or unscored) over predefined regions. The \textbf{genomic features} the package can handle can be  anything with a minimal information of chromosome,start and end. The features could have any length
and most of the time they are associated with a score. Typical examples of such data sets include aligned reads from high-throughput sequencing (HTS) experiments, percent
methylation values for CpGs (or other cytosines), locations of transcription factor binding site motifs, and so on. On the other hand, throughout the vignette we use the phrase "genomic annotation" to refer to the regions of the genome associated with a potential function and do not necessarily have a score (examples: CpG islands, genes, enhancers, promoter, exons, etc. ). These genomic annotations are usually the regions of interest, and distribution of genomic features over/around the annotations are can make the way for biological interpretation of the data.


The pipeline for computational knowledge extraction consists of three steps: data filtering, integration of data from multiple sources or generation of predictive models and biological interpretation of produced models, which leads to novel hypotheses that can be tested in the wetlab. \Rpackage{genomation} aims to facilitate the integration of multiple sources of genomic features with genomic annotation or already published experimental results.
\\*

% ----------------------------------------------------------------- %
\section{Access the data}

High-throughput data which will be used to show the functonality of the \Rpackage{genomation} is located in two places. The annotation and cap analysis of gene expression (CAGE) data comes prepared with the genomation package, while the raw HTS data can be found in the sister package \Rpackage{genomationData}.\\
To install the \Rpackage{genomation} and the complementary data package the from the github repository c/p the following lines into your R interpreter:
<<genomationDataLoad, eval=FALSE>>=
library(devtools)
install_github("genomationData",username='frenkiboy')
install_github("genomation",username='frenkiboy')
@

The \Rpackage{genomationData} vignette contains a verbose description of contained files.\\
To list the available data:
<<listGenomationData, eval=FALSE>>=
list.files(system.file('extdata',package='genomationData'))
@

To see the descriptions of the files:
<<genomationDataInfo, eval=FALSE, tidy=TRUE>>=
sampleInfo = read.table(system.file('extdata/SamplesInfo.txt',package='genomationData'),header=T, sep='\t')
head(sampleInfo)
@

Basic annotation data and processed experimental data can be found within the \Rpackage{genomation} package.
The data can be accesed throught the \Rcode{data} command or located in the extdata folder. 
<<genomationInfo, eval=FALSE, tidy=TRUE>>=
library(genomation)
data(cage)
data(cpgi)

list.files(system.file('extdata', package='genomation'))
@



% ----------------------------------------------------------------- %
\section{Data input}

One of larger hindrances in computational genomics stems from the myriad of formats
that are used to store the data. Although some formats have been selected as de facto standards
for specific kind of biological data (e.g. BAM, VCF), 
almost all publications come with supplementary tables that do not have the same structure, 
but hold similar information. The tables usually have a tabular format, contain the location
of elements in genomic coordinates and various metadata colums. 
\Rpackage(genomation) contais functions to read genomic features and genomic annotation 
provided they are in a tabular format. These functions will read the data from 
flat files into GRanges or GRangesList objects.

\Rcode{readGeneric} is the workhorse of the genomation package. It is a function
developed specifically for input of genomic data in tabular formats, and their conversion
to a GRanges object. 
By default, the function persumes that the file is a standard .bed file containing columns chr, start, end.
<<readGeneric1, eval=FALSE, tidy=TRUE>>=
library(genomation)
tab.file1=system.file("extdata/tab1.bed", package = "genomation")
readGeneric(tab.file1)
@

If the file contains meta data columns (as in extended bed format), 
it is possible to read all or some of the additional columns.
To select columns which you want to read in, use the \Rcode{meta.col} argument 
<<readGeneric2, eval=FALSE, tidy=TRUE>>=
readGeneric(tab.file1, keep.all.metadata=TRUE)

readGeneric(tab.file1, meta.col = list(CpGnum=4))
@

If the file contains header, the function will automatically recognize the 
columns using the header names.
<<readGeneric3, eval=FALSE, tidy=TRUE>>=
readGeneric(tab.file1, header=TRUE, keep.all.metadata=TRUE)
@


If the files have permutted columns, such that the first three do not represent
chromosome, start and end, you can select an arbitrary set of columns using the
chr, start and end arguments.
<<readGeneric4, eval=FALSE, tidy=TRUE>>=
tab.file2=system.file("extdata/tab2.bed", package = "genomation")
readGeneric(tab.file2, chr=3, start=2, end=1)
@

\Rcode{readGeneric]} function can easily be extended to read almost any kind of 
biological data. As an example we have provided convenience functions to read the Encode
narrowPeak and broadPeak formats, and gtf formatted files.
<<readGff1, eval=FALSE, tidy=TRUE>>=
gff.file=system.file("extdata/chr21.refseq.hg19.gtf", package = "genomation")
gff = gffToGRanges(gff.file)
head(gff)
@

In order to split the last column of the gff file, use the \Rcode{split.group=TRUE}
argument.
<<readGff2, eval=FALSE, tidy=TRUE>>=
gff = gffToGRanges(gff.file, split.group=TRUE)
head(gff)
@

There are specific functions to read genomic annotation from flat bed files.
\Rcode{readFeatureFlank} is a convenience function used to get the ranges flanking 
the set of interest. As an example, it could be used to get the CpG island shores, 
which have been shown to harbour contidion specific differential methylation
<<readFeatureFlank, eval=FALSE, tidy=TRUE>>=
# reading genes stored as a BED files
cpgi.file = system.file("extdata/chr21.CpGi.hg19.bed", package = "genomation")
cpgi.flanks = readFeatureFlank(cpgi.file)
head(cpgi.flanks$flanks)
@



<<readGeneFeatures, eval=FALSE, tidy=TRUE>>=
readGeneFeatures()
@

% ----------------------------------------------------------------- %
\section{Extraction and visualization of genomic data}

A standard step in a computational genomics experiment is the visualization of 
average enrichment over a certain predefined set of ranges, such as a visualization
of mean coverage of a certain histone modification around a transcription factor binding site,
or visualization of histone positions around transcription start sites.
\Rpackage{genomation} provides convenience functions for extraction and visualization
of data over predefined windows.

\subsection{Extration of data over predefined winows}

\Rcode{ScoreMatrix} and \Rcode{ScoreMatrixBin} are functions used to extract 
data over predefined windows. \Rcode{ScoreMatrix} is used when all of the windows
have the same width, such as an area around the transcription start site, while the
\Rcode{ScoreMatrixBin} is designed for use with windows of unequal width 
(e.g. enrichment of methylation over exons).
Both functions have 2 main arguments: \Rcode{target} and 
\Rcode{windows}. \Rcode{target} is the data that we want to extract, while the \Rcode{windows}
represents the regions over which we want to see the enrichment. The target data
can be in 3 forms: a GRanges, a RLeList or a path to an indexed .bam file.
The \Rcode{windows} must be GRanges object.

As an example we will extract the density of cage tags around the promoters on 
the human chromosome 21.
<<ScoreMatrixExample, eval=FALSE, tidy=TRUE>>=
data(cage)
data(promoters)
sm = ScoreMatrix(target=cage, windows=promoters)
sm
@


\Rcode{ScoreMatrixBin} function has an additional \Rcode{bin.num} argument which
specifies by how many bins will each window be represented (ie. it converts windows
of unequal width into ones of equal width.). By default, the binning function is set
to \Rcode{mean}.
<<ScoreMatrixBinExample, eval=FALSE, tidy=TRUE>>=
data(cage)
gff.file = system.file('extdata/chr21.refseq.hg19.gtf', package='genomation')
exons = gffToGRanges(gff.file, filter='exon')
sm = ScoreMatrixBin(target=cage, windows=exons, bin.num=50)
@

As you can see, running \Rcode{ScoreMatrixBin} with \Rcode{bin.num=50} on 
a set of exons warned us that some of the ranges had width less than 50 bp and
were removed from the set.

To simultaneously work on multiple files you can use the \Rcode{ScoreMatrixList}
function. The function likewise has 2 obligatory arguments \code{targets} and \code{windows}.
While the \code{windows} is the same as in \Rcode{ScoreMatrix}, the targets argument
is contains results from multiple experiments. 
It can be in one of the three formats: a list of RleLists, a list of GRanges (or a GRangesList object), 
or a character vector designating a set of .bam files.
<<ScoreMatrixList, eval=FALSE, tidy=TRUE>>=
data(promoters)
data(cpgi)
data(cage)

cage$tpm = NULL
targets = list(cage=cage, cpgi=cpgi)
sm = ScoreMatrixList(targets=targets, windows=promoters, bin.num=50)
@

If all of the windows have the same width \Rcode{ScoreMatrixList} will use the
\Rcode{ScoreMatrix} function. That can be overridden by explicity specifying the 
\code{bin.num} argument, as we did in the example.

\subsection{Visualization of multiple genomic experiments}

There are 2 basic modes of visualization of enrichment over windows: either as a heatmap, 
or as a line grap. \Rcode{heatMatrix]}, \Rcode{plotMeta} and \Rcode{multiHeatMatrix} are functions for visualization of \Rcode{ScoreMatrix} and \Rcode{ScoreMatrixList} objects.

We will plot the distribution of CAGE tags around promoters on human chr21.
<<heatMatrix1, eval=TRUE, tidy=TRUE, echo=FALSE, fig.width=3, fig.height=2.5, out.width='.49\\linewidth', fig.keep='last'>>=
data(cage)
data(promoters)
sm = ScoreMatrix(target=cage, windows=promoters)

par(mfrow=c(1,2))
heatMatrix(sm, xcoords=c(-1000, 1000))
plotMeta(sm, xcoords=c(-1000, 1000))
par(mfrow=c(1,1))
@


The \Rcode{heatMatrix} function can also take a list of numeric vectors designating row names, or a factor variable that represent our annotation over the windows.
<<heatMatrix2, eval=TRUE, tidy=TRUE, echo=FALSE, fig.width=3, fig.height=2.5, out.width='.49\\linewidth', fig.keep='last'>>=
data(cage)
data(promoters)
data(cpgi)

sm = ScoreMatrix(target=cage, windows=promoters, strand.aware=TRUE)
cpg.ind = which(countOverlaps(promoters, cpgi)>0)
nocpg.ind = which(countOverlaps(promoters, cpgi)==0)
heatMatrix(sm, xcoords=c(-1000, 1000), group=list(cpgi=cpg.ind, nocpg=nocpg.ind))
@

% ----------------------------------------------------------------- %
\section{Annotation of genomic features}


% ----------------------------------------------------------------- %
\section{Use cases for genomation package}
The genomation package provides generalizable functions for genomic data analysis
and visualization. Below we will demonstrate the functionality on specific use cases

% ----------------------------------------------------------------- %
\subsection{Annotation of HTS data by functional regions}


% ----------------------------------------------------------------- %
\subsection{Visualization of ChiP sequencing data}

We will visualize the binding profiles of 6 transcription factors around the 
Ctcf binding sites. 

In the fist step we will select the *.bam files containing mapped reads.
<<selectBamChipseq, eval=TRUE, tidy=TRUE, echo=FALSE>>=
library(genomationData)
library(GenomicRanges)
genomationDataPath = system.file('extdata',package='genomationData')
bam.files = list.files(genomationDataPath, full.names=T, pattern='bam$')
bam.files = bam.files[!grepl('Cage', bam.files)]
@

Firstly, we will read in the Ctcf peaks, filter regions from human chromosome 21,
and order them by their signal values. In the end we will resize all ranges to have a uniform
width of 500 bases, fixed on the center of the peak.
<<readCtcfPeaks, eval=TRUE, tidy=TRUE, echo=FALSE>>=
ctcf.peaks = readBroadPeak(file.path(genomationDataPath, 'wgEncodeBroadHistoneH1hescCtcfStdPk.broadPeak.gz'))
ctcf.peaks = ctcf.peaks[seqnames(ctcf.peaks) == 'chr21']
ctcf.peaks = ctcf.peaks[order(-ctcf.peaks$signalValue)]
ctcf.peaks = resize(ctcf.peaks, width=1000, fix='center')
@


In order to extract the coverage values of all transcription factors around 
chipseq peaks, we will use the \Rcode{ScoreMatrixList} function. 
\Rcode{ScoreMatrixList} assign names to each element of the list based on the names of the bam files. We
will use the names of the files to find the corresponding names of each sample in the SamplesInfo.txt
Using the \Rcode{heatmapProfile} on our \Rcode{ScoreMatrixList}, 
we can plot the underlying signal side by side.

\begin{figure}
<<ctcfScoreMatrixList, eval=TRUE, tidy=TRUE, echo=FALSE, fig.width=3, fig.height=2.5, out.width='.49\\linewidth', fig.keep='last'>>=
sml = ScoreMatrixList(bam.files, ctcf.peaks, bin.num=50, type='bam')
sampleInfo = read.table(system.file('extdata/SamplesInfo.txt',package='genomationData'),header=T, sep='\t')
names(sml) = sampleInfo$sampleName[match(names(sml),sampleInfo$fileName)]
multiHeatMatrix(sml, xcoords=c(-500, 500), col=c('lightgray', 'blue'))
@
\caption{Heatmap profile of unscaled coverage shows a slight colocalization of Ctcf, Rad21 and Znf143;
option \texttt{fig.keep='last'} here.\label{fig:ctcfScoreMatrixList}}
\end{figure}


Because of the large range of signal values in chipseq peaks, the \Rcode{heatmapProfile} 
will not show the true extent of colocalization. To get around this, it is advisable
to independently scale the rows of each element in the ScoreMatrixList.

\begin{figure}
<<plotScaledProfile, eval=TRUE, tidy=TRUE, echo=FALSE, fig.width=3, fig.height=2.5, out.width='.49\\linewidth', fig.keep='last'>>=
sml.scaled = scaleScoreMatrixList(sml)
multiHeatMatrix(sml.scaled, xcoords=c(-500, 500), col=c('lightgray', 'blue'))
@
\caption{Heatmap profile of scaled coverage shows much stronger colocalization of the transcription factors; nevertheless, it is evident that some of the CTCF peaks have a very weak enrichment.
option \texttt{fig.keep='last'} here.\label{fig:plotScaledProfile}}
\end{figure}

% ----------------------------------------------------------------- %
\subsection{Combinatorial binding of transcription factors}

In the first step we will read all peak files into a GRanges list. We will use the SamplesInfo file
from the \Rpackage{genomationData} to get he names of the samples. Four of the peak files are 
in the Encode broadPeak format, while one is in the narrowPeak. To read
the files, we will use the \Rcode{readGeneric} function. It enables us to select 
from the files only columns of interest. As the last step, we will restrict ourselves to
peaks that are located on chromosome 21 and have width 100 and 1000 bp
<<readAllPeaks, eval=TRUE, tidy=TRUE, echo=FALSE>>=
library(genomationData)
genomationDataPath = system.file('extdata',package='genomationData')
sampleInfo = read.table(file.path(genomationDataPath, 'SamplesInfo.txt'), header=T, sep='\t', stringsAsFactors=FALSE)


peak.files = list.files(genomationDataPath, full.names=T, pattern='Peak.gz$')
peaks = list()
for(i in 1:length(peak.files)){
    file = peak.files[i]
    name=sampleInfo$sampleName[match(basename(file),sampleInfo$fileName)]
    message(name)
    peaks[[name]] = readGeneric(file, meta.col=list(score=5))
}
peaks = GRangesList(peaks)
peaks = peaks[seqnames(peaks) == 'chr21' & width(peaks) < 1000  & width(peaks) > 100]
@

To find the combination of binding sites we will use the \Rcode{findFeatureComb} function.
It takes a granges list object, finds the union of the ranges and designates each
range by the combination of overlaps from the original set.
By default, the returned ranges will have a numeric \Rcode{class} meta data column, which designates
the correponding combination. If you are interested in the names of the TF which make 
the combinations, put the \Rcode{use.names=TRUE}.
<<findFeatureComb, eval=TRUE, tidy=TRUE, echo=FALSE>>=
tf.comb = findFeatureComb(peaks, width=1000)
@


To visualize the results, we will plot the enricment of resulting regions. Before doing that we will
order the regions by their \Rcode{class} argument.
\begin{figure}
<<visualizeFeatureComb, eval=TRUE, tidy=TRUE, echo=FALSE, fig.width=3, fig.height=2.5, out.width='.49\\linewidth', fig.keep='last'>>=
tf.comb = tf.comb[order(tf.comb$class)]
bam.files = list.files(genomationDataPath, full.names=T, pattern='bam$')
bam.files = bam.files[!grepl('Cage', bam.files)]
sml = ScoreMatrixList(bam.files, tf.comb, bin.num=20, type='bam')
names(sml) = sampleInfo$sampleName[match(names(sml),sampleInfo$fileName)]
sml.scaled = scaleScoreMatrixList(sml)
multiHeatMatrix(sml.scaled, xcoords=c(-500, 500), col=c('lightgray', 'blue'))
@
\end{figure}
The plot shows perfectly how misleading the peak calling process can be. 
Although the plots show that CTFC, Rad21 and Znf143 have almost perfect colozalization,
Peak callers have trouble identifying peaks in regions with lower enrichments and as a 
result, we get different statistics when using overlaps.

% ----------------------------------------------------------------- %


\subsection{Annotation of bam files}


\end{document}


